src/directives/run-async.ts

```diff
@@ -0,0 +1,153 @@
+    // Abort a pending request if there is one
+    if (currentRunState !== undefined && currentRunState.state === 'pending') {
+      if (currentRunState.abortController !== undefined) {
+        currentRunState.abortController.abort();
+      }
+      // TODO(justinfagnani): This should be an AbortError, but it's not
+      // implemented yet
+      currentRunState.rejectPending(new Error());
+    }
+    const abortController =
+        hasAbortController ? new AbortController() : undefined;
+    const abortSignal =
+        hasAbortController ? abortController!.signal : undefined;
+    let resolvePending!: () => void;
+    let rejectPending!: (e: Error) => void;
+    const pendingPromise = new Promise((res, rej) => {
+      resolvePending = res;
+      rejectPending = rej;
+    });
+    const promise = task(key, {signal: abortSignal});
+    // The state is immediately 'pending', since the function has been
+    // executed, but if the function throws an InitialStateError to
+    // indicate that it couldn't even start processing, then we will set
+    // the state to 'initial'.
+    const runState: AsyncRunState = {
+      key,
+      promise,
+      state: 'pending',
+      abortController,
+      resolvePending,
+      rejectPending,
+    };
+    runs.set(part, runState);
+
+    Promise.resolve(promise).then(
+        (value: unknown) => {
+          runState.state = 'success';
+          const currentRunState = runs.get(part);
+          runState.resolvePending();
+          if (currentRunState !== runState) {
+            return;
+          }
+          part.setValue(success(value));
+          part.commit();
+        },
+        (error: Error) => {
+          const currentRunState = runs.get(part);
+          runState.rejectPending(new Error());
+          if (currentRunState !== runState) {
+            return;
+          }
+          if (error instanceof InitialStateError &&
+              typeof initial === 'function') {
+            runState!.state = 'initial';
+            part.setValue(initial());
+            part.commit();
+          } else {
+            runState!.state = 'failure';
+            if (typeof failure === 'function') {
+              // render success callback
+              part.setValue(failure(error));
+              part.commit();
+            }
+          }
+        });
+
+    (async () => {
+      // Wait a microtask for the initial render of the Part to complete
+      await 0;
+      const currentRunState = runs.get(part);
+      if (currentRunState === runState && currentRunState.state === 'pending') {
+        part.startNode.parentNode!.dispatchEvent(
+            new CustomEvent('pending-state', {
+              composed: true,
+              bubbles: true,
+              detail: {promise: pendingPromise}
+            }));
+      }
+    })();
+  }
+
+  // If the promise has not yet resolved, set/update the defaultContent
+  if ((currentRunState === undefined || currentRunState.state === 'pending') &&
```

Comment:
Hi @justinfagnani : I am experimenting with this a bit since you mentioned it last week.  I notice that the pending state is not applied on first key/dependency change after a previous success state - it remains in the success state.  Another key change is required for the pending state to be applied.  

I could be doing something incorrectly, but I think : on line#93 `runs.set(part, runState);` is called to set a new pending state, but `currentRunState` is still the previous success state, so `pending()` is not called here.  Once the directive is run again (ex. second key-stroke in the search demo), `currentRunState` is initialized with the most recent pending state, and then `pending()` is called here.  It is perhaps not super noticeable when there are many changes (like `input` event).  The absent intermediate pending state was fairly obvious in my case where there aren't multiple key changes quickly in succession.

Reply:
I would agree with this. I wonder what the desired benefits of this test being anything more than `if (typeof pending === 'function') {`? I've been playing with the code shaped this way at https://stackblitz.com/edit/lit-element-run-async-2?file=run-async.ts